The exists method lacks permission check

**Alluxio Version:**
master.

**Describe the bug**

I tested the exists method of HDFS and Alluxio. HDFS need to have execution permission on the parent folder and Alluxio has no permission to check.

**To Reproduce**
### HDFS
test code:
```
public static void printStatus(FileSystem fileSystem, String path) {
    try {
        System.out.println(fileSystem.exists(new Path("/a/b/c")));
    } catch (Exception e) {
        e.printStackTrace();
    }
}
```

- without parent permission:

```
drwx------   - alluxio staff          0 2022-08-04 15:41 /a/b
```

output:
```
Caused by: org.apache.hadoop.ipc.RemoteException(org.apache.hadoop.security.AccessControlException): Permission denied: user=hdfs, access=EXECUTE, inode="/a/b/c":hdfs:staff:drwx------
```

- with parent permission:

```
drwx--x--x   - alluxio staff          0 2022-08-04 15:41 /a/b
```

output:
```
true
```

### Alluxio

test code:
```
public static void exists() throws Exception {
   FileSystem fs = getFileSystem();
   System.out.println(fs.exists(new AlluxioURI("/a/b/c")));
}
```

- without parent permission:
```
drwx------  alluxio    staff                        2       PERSISTED 08-31-2022 09:05:59:745  DIR /a/b
```

output:
```
true
```

**Expected behavior**
The exists method requires permission check.

**Urgency**
Urgency.

**Are you planning to fix it**
Yes.

--- a/core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java
+++ b/core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java
@@ -1471,7 +1471,7 @@ public class DefaultFileSystemMaster extends CoreMaster
           loadMetadataIfNotExist(rpcContext, path, lmCtx, false);
         }
       } catch (FileDoesNotExistException e) {
-        return false;
+        // ignore exception
       }
 
       try (LockedInodePath inodePath = mInodeTree.lockInodePath(

--- a/core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java
+++ b/core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java
@@ -1462,12 +1462,6 @@ public class DefaultFileSystemMaster extends CoreMaster
           createLockingScheme(path, context.getOptions().getCommonOptions(), LockPattern.READ),
           rpcContext.getJournalContext())
       ) {
-        try {
-          mPermissionChecker.checkParentPermission(Mode.Bits.EXECUTE, inodePath);
-        } catch (AccessControlException e) {
-          auditContext.setAllowed(false);
-          throw e;
-        }
         LoadMetadataContext lmCtx = LoadMetadataContext.create(
             LoadMetadataPOptions.newBuilder()
                 .setCommonOptions(context.getOptions().getCommonOptions())
@@ -1485,7 +1479,9 @@ public class DefaultFileSystemMaster extends CoreMaster
           rpcContext.getJournalContext())
       ) {
         try {
-          mPermissionChecker.checkParentPermission(Mode.Bits.EXECUTE, inodePath);
+          if (mPermissionChecker instanceof DefaultPermissionChecker) {
+            mPermissionChecker.checkParentPermission(Mode.Bits.EXECUTE, inodePath);
+          }
         } catch (AccessControlException e) {
           auditContext.setAllowed(false);
           throw e;

--- a/core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java
+++ b/core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java
@@ -1462,6 +1462,12 @@ public class DefaultFileSystemMaster extends CoreMaster
           createLockingScheme(path, context.getOptions().getCommonOptions(), LockPattern.READ),
           rpcContext.getJournalContext())
       ) {
+        try {
+          mPermissionChecker.checkParentPermission(Mode.Bits.EXECUTE, inodePath);
+        } catch (AccessControlException e) {
+          auditContext.setAllowed(false);
+          throw e;
+        }
         LoadMetadataContext lmCtx = LoadMetadataContext.create(
             LoadMetadataPOptions.newBuilder()
                 .setCommonOptions(context.getOptions().getCommonOptions())
@@ -1478,6 +1484,12 @@ public class DefaultFileSystemMaster extends CoreMaster
           createLockingScheme(path, context.getOptions().getCommonOptions(), LockPattern.READ),
           rpcContext.getJournalContext())
       ) {
+        try {
+          mPermissionChecker.checkParentPermission(Mode.Bits.EXECUTE, inodePath);
+        } catch (AccessControlException e) {
+          auditContext.setAllowed(false);
+          throw e;
+        }
         auditContext.setSucceeded(true);
         return inodePath.fullPathExists();
       }

--- a/core/server/master/src/test/java/alluxio/master/file/FileSystemMasterFsOptsTest.java
+++ b/core/server/master/src/test/java/alluxio/master/file/FileSystemMasterFsOptsTest.java
@@ -45,6 +45,7 @@ import alluxio.grpc.SetAttributePOptions;
 import alluxio.master.file.contexts.CreateDirectoryContext;
 import alluxio.master.file.contexts.CreateFileContext;
 import alluxio.master.file.contexts.DeleteContext;
+import alluxio.master.file.contexts.ExistsContext;
 import alluxio.master.file.contexts.GetStatusContext;
 import alluxio.master.file.contexts.ListStatusContext;
 import alluxio.master.file.contexts.MountContext;
@@ -1281,6 +1282,33 @@ public class FileSystemMasterFsOptsTest extends FileSystemMasterTestBase {
     assertEquals(files * 3 + 3, infos.size());
     for (FileInfo info : infos) {
       assertEquals(newEntries, Sets.newHashSet(info.convertAclToStringEntries()));
+    }
+  }
+
+  @Test
+  public void exists() throws Exception {
+    createFileWithSingleBlock(NESTED_FILE_URI, mNestedFileContext);
+
+    //Test existing file
+    assertEquals(true, mFileSystemMaster.exists(NESTED_FILE_URI, ExistsContext.defaults()));
+
+    //Test non-existent file
+    assertEquals(false, mFileSystemMaster.exists(NESTED_FILE2_URI, ExistsContext.defaults()));
+
+    //Test file without parent permission
+    mFileSystemMaster.setAttribute(NESTED_URI, SetAttributeContext
+        .mergeFrom(SetAttributePOptions.newBuilder().setMode(new Mode((short) 0700).toProto())));
+    try (AuthenticatedClientUserResource userA = new AuthenticatedClientUserResource("userA",
+        Configuration.global())) {
+      mThrown.expect(AccessControlException.class);
+      mFileSystemMaster.exists(NESTED_FILE_URI, ExistsContext.defaults());
+    }
+
+    //Test non-existent file without parent permission
+    try (AuthenticatedClientUserResource userA = new AuthenticatedClientUserResource("userA",
+        Configuration.global())) {
+      mThrown.expect(AccessControlException.class);
+      mFileSystemMaster.exists(NESTED_FILE2_URI, ExistsContext.defaults());
     }
   }
 }

